datasource/src/main/scala/quasar/physical/mongo/MongoDatasource.scala

```diff
@@ -0,0 +1,95 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+import quasar.api.datasource.DatasourceType
+import quasar.api.resource._
+import quasar.connector.{MonadResourceErr, QueryResult, ResourceError}
+import quasar.connector.datasource._
+import quasar.physical.mongo.decoder._
+import quasar.physical.mongo.MongoResource.{Database, Collection}
+
+import cats.syntax.eq._
+import cats.syntax.functor._
+import cats.effect._
+import cats.syntax.applicative._
+import cats.syntax.option._
+import eu.timepit.refined.auto._
+import fs2.Stream
+import shims._
+
+class MongoDataSource[F[_]: ConcurrentEffect: ContextShift: MonadResourceErr: Timer](mongo: Mongo[F])
+  extends LightweightDatasource[F, Stream[F, ?], QueryResult[F]] {
+
+  val kind = MongoDataSource.kind
+
+  override def evaluate(path: ResourcePath): F[QueryResult[F]] = {
+    val errorStream =
+      Stream.raiseError(ResourceError.throwableP(ResourceError.pathNotFound(path)))
+
+    val stream = path match {
+      case _ if isRoot(path) => errorStream
+      case ResourcePath.Leaf(file) => MongoResource.ofFile(file) match {
+        case None => errorStream
+        case Some(Database(_)) => errorStream
+        case Some(collection@Collection(_, _)) => mongo.findAll(collection)
+      }
+    }
+    QueryResult.parsed(qdataDecoder, stream).pure[F]
+  }
+
+  override def pathIsResource(path: ResourcePath): F[Boolean] = path match {
+    case _ if isRoot(path) => false.pure[F]
+    case ResourcePath.Leaf(file) => MongoResource.ofFile(file) match {
+      case Some(Database(_)) => false.pure[F]
+      case Some(coll@Collection(_, _)) => mongo.collectionExists(coll).compile.lastOrError
+      case None => false.pure[F]
+    }
+  }
+
+  override def prefixedChildPaths(prefixPath: ResourcePath)
+    : F[Option[Stream[F, (ResourceName, ResourcePathType)]]] = prefixPath match {
+    case _ if isRoot(prefixPath) =>
+      mongo.databases.map(x => (ResourceName(x.name), ResourcePathType.prefix)).some.pure[F]
+
+    case ResourcePath.Leaf(file) => MongoResource.ofFile(file) match {
+      case None => none.pure[F]
+      case Some(Collection(_, _)) => none.pure[F]
+      case Some(db@Database(_)) => {
+        val dbExists: F[Boolean] = mongo.databaseExists(db).compile.last.map(_.getOrElse(false))
+
+        dbExists.map(exists => {
+          if (exists) {
+            mongo.collections(db).map(x => (ResourceName(x.name), ResourcePathType.leafResource)).some
+          } else {
+            none
+          }
+        })
+      }
+    }
+  }
+  private def isRoot(inp: ResourcePath): Boolean =
+    inp === (ResourcePath.root() / ResourceName("")) || inp === ResourcePath.root()
```

Comment:
Why are we special casing `ResourcePath.root() / ResourceName("")` here?

Reply:
`ResourcePath.root() / ResourceName("")` looks like path encoding edge case, that is for some reasone has special case in s3 and azure connectors. If `/` could be parsed as `ResourcePath.root()` and `ResourcePath.root() / ResourceName("")` then it's necessary to handle both as the same thing, otherwise I'm happily remove this strange case out of the code 🙂

Reply:
Yeah, I'd remove it. If we're getting that as input from somewhere (sdbe?) then we should fix it there instead of special-casing everywhere.

Reply:
We probably ought to fix that in `quasar` rather than special casing in all connectors.

Reply:
Yeah same reason as https://github.com/slamdata/quasar-datasource-azure/pull/19

If we remove it here, then SDBE integration won't work at all (repl integration would) so that would mean it needs to be fixed soon. BTW I don't really care if we leave it in for now or remove it, just mentioning it to be explicit.